src/pkg/retention/res/selectors/regexp/selector_test.go

```diff
@@ -0,0 +1,159 @@
+	suite.Suite
+
+	artifacts []*res.Candidate
+}
+
+// TestRegExpSelector is entrance for RegExpSelectorTestSuite
+func TestRegExpSelector(t *testing.T) {
+	suite.Run(t, new(RegExpSelectorTestSuite))
+}
+
+// SetupSuite to do preparation work
+func (suite *RegExpSelectorTestSuite) SetupSuite() {
+	suite.artifacts = []*res.Candidate{
+		{
+			NamespaceID:  1,
+			Namespace:    "library",
+			Repository:   "harbor",
+			Tag:          "latest",
+			Kind:         res.Image,
+			PushedTime:   time.Now().Unix() - 3600,
+			PulledTime:   time.Now().Unix(),
+			CreationTime: time.Now().Unix() - 7200,
+			Labels:       []string{"label1", "label2", "label3"},
+		},
+		{
+			NamespaceID:  1,
+			Namespace:    "library",
+			Repository:   "redis",
+			Tag:          "4.0",
+			Kind:         res.Image,
+			PushedTime:   time.Now().Unix() - 3600,
+			PulledTime:   time.Now().Unix(),
+			CreationTime: time.Now().Unix() - 7200,
+			Labels:       []string{"label1", "label4", "label5"},
+		},
+		{
+			NamespaceID:  1,
+			Namespace:    "library",
+			Repository:   "redis",
+			Tag:          "4.1",
+			Kind:         res.Image,
+			PushedTime:   time.Now().Unix() - 3600,
+			PulledTime:   time.Now().Unix(),
+			CreationTime: time.Now().Unix() - 7200,
+			Labels:       []string{"label1", "label4", "label5"},
+		},
+	}
+}
+
+// TestTagMatches tests the tag `matches` case
+func (suite *RegExpSelectorTestSuite) TestTagMatches() {
+	tagMatches := &selector{
+		decoration: Matches,
+		pattern:    "{latest,4.*}",
+	}
+
+	selected, err := tagMatches.Select(suite.artifacts)
+	require.NoError(suite.T(), err)
+	assert.Equal(suite.T(), 3, len(selected))
+
+	tagMatches2 := &selector{
+		decoration: Matches,
+		pattern:    "4.*",
+	}
+
+	selected, err = tagMatches2.Select(suite.artifacts)
+	require.NoError(suite.T(), err)
+	assert.Equal(suite.T(), 2, len(selected))
+}
+
+// TestTagExcludes tests the tag `excludes` case
+func (suite *RegExpSelectorTestSuite) TestTagExcludes() {
+	tagExcludes := &selector{
+		decoration: Excludes,
+		pattern:    "{latest,4.*}",
+	}
+
+	selected, err := tagExcludes.Select(suite.artifacts)
+	require.NoError(suite.T(), err)
+	assert.Equal(suite.T(), 0, len(selected))
+
+	tagExcludes2 := &selector{
+		decoration: Excludes,
+		pattern:    "4.*",
+	}
+
+	selected, err = tagExcludes2.Select(suite.artifacts)
+	require.NoError(suite.T(), err)
+	assert.Equal(suite.T(), 1, len(selected))
+}
+
+// TestRepoMatches tests the repository `matches` case
+func (suite *RegExpSelectorTestSuite) TestRepoMatches() {
+	repoMatches := &selector{
+		decoration: RepoMatches,
+		pattern:    "{redis}",
+	}
+
+	selected, err := repoMatches.Select(suite.artifacts)
+	require.NoError(suite.T(), err)
+	assert.Equal(suite.T(), 2, len(selected))
+
+	repoMatches2 := &selector{
+		decoration: RepoMatches,
+		pattern:    "har*",
+	}
+
+	selected, err = repoMatches2.Select(suite.artifacts)
+	require.NoError(suite.T(), err)
+	assert.Equal(suite.T(), 1, len(selected))
```

Comment:
Same as the other test, we should probably check that we're matching the right thing.

Reply:
Sure